src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -245,6 +246,7 @@ internal bool TryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions opt
             if (ClassType == ClassType.Value)
             {
                 Debug.Assert(!state.IsContinuation);
+                Debug.Assert(value != null);
```

Comment:
This `Debug.Assert` will fail in this case. This is similar to the bug: https://github.com/dotnet/runtime/issues/32523

```C#
[Fact]
public static void WriteArrayOfStringsWithNull()
{
    string[] strs = new string[] { "foo", null };
    string json = JsonSerializer.Serialize(strs, new JsonSerializerOptions { Converters = { new StringConverter() } });
}

public class StringConverter : JsonConverter<string>
{
    public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }

    public override void Write(Utf8JsonWriter writer, [DisallowNull] string value, JsonSerializerOptions options)
    {
        if (value == null)
        {
            throw new InvalidOperationException("The custom converter should never get called with null value.");
        }
        writer.WriteStringValue(value);
    }
}
```

Reply:
cc @steveharter, @layomia  - fyi

Reply:
Should we remove the assert then and leave the bang for now? when the bug is fixed a null comparison should be done and special case it in another block?

Reply:
Yep, sounds good.